site/content/blog/kapp-versioned-resource.md

```diff
@@ -0,0 +1,566 @@
+
+Continue? [yN]: y
+```
+
+Let's check the value of environment variable `MSG_KEY` in the running pod of deployment `simple-app`.
+```bash
+$ kubectl get pods            
+NAME                        READY   STATUS    RESTARTS   AGE
+simple-app-657f9c8494-t2pw9   1/1     Running   0          99s
+
+$ kubectl exec -it simple-app-657f9c8494-t2pw9 sh
+# echo $MSG_KEY
+hello-carvel
+# exit
+#
+```
+
+Let's update the value of `data.hello_msg ` to `hello-kapp` in configmap `simple-config` and re-deploy the app:
+```bash
+$ kapp deploy -a app -f app.yaml --diff-changes
+Target cluster 'https://127.0.0.1:33907' (nodes: minikube)
+
+@@ update configmap/simple-config (v1) namespace: default @@
+  ...
+  1,  1   data:
+  2     -   hello_msg: hello-carvel
+      2 +   hello_msg: hello-kapp
+  3,  3   kind: ConfigMap
+  4,  4   metadata:
+
+Changes
+
+Namespace  Name           Kind       Age  Op      Op st.  Wait to    Rs  Ri  
+default    simple-config  ConfigMap  3d   update  -       reconcile  ok  -  
+
+Op:      0 create, 0 delete, 1 update, 0 noop, 0 exists
+Wait to: 1 reconcile, 0 delete, 0 noop
+
+Continue? [yN]: y
+```
+
+Now let's verify again the value of environment variable `MSG_KEY` in the running pod of deployment `simple-app`.
+
+```bash
+$ kubectl get pods            
+NAME                        READY   STATUS    RESTARTS   AGE
+simple-app-657f9c8494-t2pw9   1/1     Running   0          6m40s
+
+$ kubectl exec -it simple-app-657f9c8494-t2pw9 sh
+# echo $MSG_KEY
+hello-carvel
+# exit
+#
+```
+Here, the value of environment variable `MSG_KEY` is still not updated. To reflect the new changes of configmap we have to re-start the pod manually.
+
+```bash
+$ kubectl delete pod simple-app-657f9c8494-t2pw9 
+pod "simple-app-657f9c8494-t2pw9" deleted
+
+$ kubectl get pod 
+NAME                          READY   STATUS    RESTARTS   AGE
+simple-app-797ff748db-mqx97   1/1     Running   0          6s
+
+$ kubectl exec -it simple-app-797ff748db-mqx97 sh
+# echo $MSG_KEY
+hello-kapp
+# 
+```
+After restarting the pod we can see the new changes we made in configmap.
+
+In above example, we saw that to reflect the changes of a configmap we need to restart the pod manually.
+
+## Versioned resource in `kapp`:
+
+kapp has a concept of [versioned resources](https://carvel.dev/kapp/docs/v0.49.0/diff/#versioned-resources), where it creates a new version for resource whenever a change is made to it. To enable versioning we just need to add the annotation `kapp.k14s.io/versioned: ""` to the resource. Resources which are using this annotaion will follow the naming convention `{resource-name}-ver-{n}` where `n` will start with `1` and will get increamented by `1` on every update.
+
+Whenever we make a change to a resource marked as versioned, entirely new resource will get created by the kapp instead of updating the existing one. Also, it will update the new name to the referencing resource and re-start them to reflect the new changes.
```

Comment:
```suggestion
Whenever we make a change to a resource marked as versioned, an entirely new resource will get created by kapp instead of updating the existing one. Also, it will update the new name to the referencing resource and re-start them to reflect the new changes.
```